python_interface/design/02_pace_python_implementation_discussion.md

```diff
@@ -0,0 +1,349 @@
+The only advantage of the C++ API compared to the C API is that the C++ API allows a `mex` file to get/set the properties of a Matlab object which the C API does not.
+
+There may thus be a case for using a C API `mex` file to get the reference to an engine and then use a C++ API in the actual Python module.
+However, how this is best done is left as work for the future.
+
+
+# <a name="app_memcopy"></a> Appendix: Code demonstrating the memory copies in converting data from Matlab to Python.
+
+First we need a mex file which will print the address of the actual memory location of the *data* (not the `mxArray` wrapper) of a particular Matlab array.
+This is provided by the [`get_address.mex`](https://github.com/mducle/hugo/blob/master/src/get_address.cpp) file in the Hugo repository.
+Then we can use the following Python script to get the addresses in memory of the data associated with a single array as it is created in Python, exported to Matlab and re-exported back to Python:
+
+```python
+import struct
+import ctypes
+
+from pyHorace import Matlab
+m = Matlab()
+
+# We cannot directly create a variable in the Matlab base namespace
+# So we first create it in Python using the default matlab.double ctor and get its memory address
+import matlab
+mm = matlab.double([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+addr_py = mm._data.buffer_info()[0]
+print('Original Matlab array in Python memory address is {}'.format(addr_py))
+
+# Check that it contains the correct data (in column-major, so: [1, 4, 7, 2, 5, 8, 3, 6, 9])
+data_orig = [struct.unpack('<d', ctypes.string_at(addr_py+i*8, 8))[0] for i in range(9)]
+print(data_orig)
+
+# Now we use the get_address mex file to find the address of the matrix in Matlab
+addr_mat = int(m.interface.call('get_address', [mm]))
+print('Matlab array memory address is {}'.format(addr_mat))
+
+# Luckily if we assign this variable to the Matlab base workspace, it _doesn't_ make another copy
+m.interface.call('assignin', ['base', 'mm_mat', mm])
+m.interface.call('evalin', ['base', 'whos'], nargout=0)
+addr_base = int(m.interface.call('evalin', ['base', 'get_address(mm_mat)']))
+print('Matlab base workspace array memory address is {}'.format(addr_base))
+
+# Unfortunately, if we re-export this back to Python it creates another data copy.
+# We use evalin (because we cannot use assignin to export to Python) 
+mm_py = m.interface.call('evalin', ['base', 'mm_mat'])
+print(type(mm_py))
+print(mm_py)
+
+# The data is stored in the _data attribute as a buffer object
+print(type(mm_py._data))
+print(mm_py._data)
+# The address of the buffer can be obtained from the buffer_info property
+# The first element is the address, the second is the length
+buf_info = mm_py._data.buffer_info()
+print(buf_info)
+print('Exported Python array memory address is {}'.format(buf_info[0]))
+
+# Check that the memory pointed to by the buffer info stores the right data
+# Assumming the data is stored as IEEE double, 64-bit (8 byte) floating point number
```

Comment:
Herbert/Horace is using a mix of IEEE float (32-bit) and doubles (64-bit) for data. Specifically the pixel data signal/error data is stored as 32-bit floats (@abuts may be able to add more specific details).
This interface between Python and MATLAB will need to be able to handle both float and double data.

Reply:
It should be able to do this - the interface `mex` functions will use templates to satisfy the full list of types. I just needed to assume one type for the `struct.unpack` call in the next line (since this is Python looking at raw memory)